src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs

```diff
@@ -8,19 +8,27 @@
 using Microsoft.CodeAnalysis.Experiments;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.CodeAnalysis.Shared.TestHooks;
-using Microsoft.VisualStudio.LanguageServices.Experimentation;
+using Microsoft.VisualStudio.OperationProgress;
 using Microsoft.VisualStudio.Shell;
-using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation
 {
     [ExportWorkspaceServiceFactory(typeof(IWorkspaceStatusService), ServiceLayer.Host), Shared]
     internal class VisualStudioWorkspaceStatusServiceFactory : IWorkspaceServiceFactory
     {
+        private IAsyncServiceProvider _serviceProvider;
+
+        [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioWorkspaceStatusServiceFactory()
+        public VisualStudioWorkspaceStatusServiceFactory(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
         {
+            threadingContext.JoinableTaskFactory.Run(async () =>
```

Comment:
@AArnott how am I supposed to get IAsyncServiceProvider through MEF in mef v2?

Reply:
https://github.com/dotnet/roslyn/blob/defeca5d0e44915fc978cfd93d3de2e4a6d0dbf4/src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs#L56

is what we're doing in other places.

Reply:
Yes, like Jason said. Don't have thread affinity in your importing constructors, please.

Reply:
@jasonmalinowski @AArnott 
 [Import(typeof(SVsServiceProvider))]

this doesn't work with MEF v2. Import attribute with typeof doesn't exist for v2. if I give it contract name (String) of the fully qualified name of SVSServiceProvider, it looks like it doesn't get picked up?

but if that is a right way to do it, I can try again.

Reply:
@AArnott 
> Yes, like Jason said. Don't have thread affinity in your importing constructors, please.

casting SVSServiceProvider to IVsXXXX won't cause QI? or is it free-threaded QI for SVSServiceProvider?

Reply:
It's a managed type in the container AFAIK and thus there's no QI automatically.

@heejaechang Where are you using MEF v2? I would have assumed everything moved to V3 at this point that's in the VS layer where you were using these interfaces?

Reply:
Yes given the MEFv2 attribute constraint, just use `[Import] SVsServiceProvider serviceProvider` and then cast to `IAsyncServiceProvider`. This is a free-threaded object so you can cast freely. That's all MEF was doing for you with the MEFv1 attribute anyway.

Reply:
@jasonmalinowski There's a distinction between MEF engines and MEF attributes. There are no V3 attributes. But the V3 engine supports both V1 and V2 attributes. 

But @heejaechang I see you're in fact using V1 attributes (`using System.ComponentModel.Composition;`) so you should in fact have the `ImportAttribute(Type)` overload.

Reply:
Yeah that's what I meant but said poorly. For *engines* Roslyn unit tests all use V3, so I'm surprised @heejaechang had problems. And the V1 attributes are used in the VS layer (along with V2 ones when we need to.)

Reply:
@AArnott @jasonmalinowski    it is using using System.Composition; not System.ComponentModel.Composition.

Reply:
> I think you are confused with other files that actually use v1.

Indeed. I was looking at the file @jasonmalinowski linked to instead of to yours. 
So ya, just import SVsServiceProvider then cast it to IAsyncServiceProvider and you should be good. You can do this on any thread since the global service provider obtained from MEF is a free-threaded object.